File: dotnet/SemaphoreFifoED/SemaphoreFifoED.cs


using SynchUtils;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;


namespace SemaphoreFifoED
{
    public class SemaphoreFifoED
    {
        private object monitor;
        private int permits;

        private class Request {
            public readonly int units;
            
            public Request(int units) {
                this.units = units;
            }

            public bool Processed {
                get; set;
            }
        }

        private LinkedList<Request> requests;


        private void notifyWaiters() {
            int total = 0;
            do {
                if (requests.Count == 0) break;
                LinkedListNode<Request> first = requests.First;
                if (permits < first.Value.units) break;
                permits -= first.Value.units;
                requests.RemoveFirst();
            }
            while (true);
            if (total > 0) Monitor.PulseAll(monitor);
        }

        public SemaphoreFifoED(int initialPermits) {
            monitor = new object();
            if (initialPermits > 0) permits = initialPermits;
            requests = new LinkedList<Request>();
        }

        public bool Acquire(int units, int timeout) {
            lock(monitor) {
                // non blocking path
                if (requests.Count == 0 && permits >= units) {
                    permits -= units;
                    return true;
                }

                // try path
                if (timeout == 0) return false;

                // prepare wait
                TimeoutHolder th = new TimeoutHolder(timeout);
                LinkedListNode<Request> node = requests.AddLast(new Request(units));

                do {
                    try {
                        Monitor.Wait(monitor, th.Remaining);
                        if (node.Value.Processed) return true;
                        if (th.Timeout) {
                            requests.Remove(node);
                            notifyWaiters();
                            return false;
                        }
                    }
                    catch(ThreadInterruptedException) {
                        if (node.Value.Processed) {
                            Thread.CurrentThread.Interrupt();
                            return true;
                        }
                        requests.Remove(node);
                        notifyWaiters();
                        throw;
                    }
                }
                while (true);
            
            }
        }

        public void Release(int units) {
            lock(monitor) {
                permits += units;
                notifyWaiters();
            }
        }
    }
}

File: java/aula05-monitors/src/main/java/pt/isel/pc/SemaphoreFifoED.java
package pt.isel.pc;

import pt.isel.pc.utils.NodeList;
import pt.isel.pc.utils.TimeoutHolder;

public class SemaphoreFifoED {
    private int permits;
    private Object monitor = new Object();

    private static class Request {
        public final int units;
        private boolean done;

        public Request(int units) {
            this.units = units;
        }
        public void complete() {
            this.done = true;
        }
        public boolean isCompleted() {
            return done;
        }
    }

    private NodeList<Request> requests;

    private void notifyWaiters() {
        boolean toNotify = false;
        while ( !requests.empty() && permits >= requests.first().units) {
            Request req = requests.removeFirst();
            permits -= req.units;
            req.complete();
            toNotify = true;
        }
        if (toNotify)
            monitor.notifyAll();
    }

    public SemaphoreFifoED(int initialUnits) {
        if (initialUnits > 0)
            permits = initialUnits;
        requests = new NodeList<>();
    }

    public boolean acquire(int units, long timeout)
            throws InterruptedException {
        synchronized (monitor) {
            // non blocking path
            if (requests.empty() && permits >= units) {
                permits -= units;
                return true;
            }
            if (timeout == 0) {
                return false;
            }

            TimeoutHolder th = new TimeoutHolder(timeout);
            Request req = new Request(units);
            NodeList.Node<Request> node = requests.addLast(req);
            do {
                try {
                    monitor.wait(th.remaining());
                    if (req.isCompleted()) {
                        return true;
                    }
                    if (th.timeout()) {
                        requests.remove(node);
                        notifyWaiters();
                        return false;
                    }
                } catch (InterruptedException e) {
                    if( req.isCompleted()) {
                        // delay the interruption and return success
                        Thread.currentThread().interrupt();
                        return true;
                    }
                    requests.remove(node);
                    notifyWaiters();
                    throw e;
                }

            } while (true);
        }
    }

    public boolean acquire(int units)
            throws InterruptedException {
        return acquire(units, TimeoutHolder.INFINITE);
    }

    public void release(int units) {
        synchronized(monitor) {
            permits += units;
            notifyWaiters();
        }
    }
}
